sidecar/internal/service/networkfence.go

```diff
@@ -0,0 +1,117 @@
+	"k8s.io/klog/v2"
+)
+
+// NetworkFenceServer struct of sidecar with supported methods of proto
+// networkFence server spec and controller client to csi driver.
+type NetworkFenceServer struct {
+	proto.UnimplementedNetworkFenceServer
+	controllerClient fence.FenceControllerClient
+	kubeClient       *kubernetes.Clientset
+}
+
+// NewNetworkFenceServer creates a new NetworkFenceServer which handles the proto.NetworkFence
+// Service requests.
+func NewNetworkFenceServer(c *grpc.ClientConn, kc *kubernetes.Clientset) *NetworkFenceServer {
+	return &NetworkFenceServer{
+		controllerClient: fence.NewFenceControllerClient(c),
+		kubeClient:       kc,
+	}
+}
+
+// RegisterService registers service with the server.
+func (ns *NetworkFenceServer) RegisterService(server grpc.ServiceRegistrar) {
+	proto.RegisterNetworkFenceServer(server, ns)
+}
+
+// FenceClusterNetwork fetches required information from kubernetes cluster and calls
+// CSI-Addons FenceClusterNetwork service.
+func (ns *NetworkFenceServer) FenceClusterNetwork(
+	ctx context.Context,
+	req *proto.NetworkFenceRequest) (*proto.NetworkFenceResponse, error) {
+	// Get the secrets from the k8s cluster
+	data, err := kube.GetSecret(ctx, ns.kubeClient, req.GetSecretName(), req.GetSecretNamespace())
+	if err != nil {
+		klog.Errorf("Failed to get secret %s in namespace %s: %v", req.GetSecretName(), req.GetSecretNamespace(), err)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	cidr := req.GetCidrs()
+	fenceRequest := fence.FenceClusterNetworkRequest{
+		Parameters: req.GetParameters(),
+		Cidrs:      getCIDRS(cidr),
+		Secrets:    data,
+	}
+
+	_, err = ns.controllerClient.FenceClusterNetwork(ctx, &fenceRequest)
+	if err != nil {
+		klog.Errorf("Failed to fence cluster network: %v", err)
+		return nil, err
+	}
+
+	return &proto.NetworkFenceResponse{}, nil
+}
+
+// UnFenceClusterNetwork fetches required information from kubernetes cluster and calls
+// CSI-Addons UnFenceClusterNetwork service.
+func (ns *NetworkFenceServer) UnFenceClusterNetwork(
+	ctx context.Context,
+	req *proto.NetworkFenceRequest) (*proto.NetworkFenceResponse, error) {
+	// Get the secrets from the k8s cluster
+	data, err := kube.GetSecret(ctx, ns.kubeClient, req.GetSecretName(), req.GetSecretNamespace())
+	if err != nil {
+		klog.Errorf("Failed to get secret %s in namespace %s: %v", req.GetSecretName(), req.GetSecretNamespace(), err)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	cidr := req.GetCidrs()
+	fenceRequest := fence.UnfenceClusterNetworkRequest{
+		Parameters: req.GetParameters(),
+		Cidrs:      getCIDRS(cidr),
+		Secrets:    data,
+	}
+
+	_, err = ns.controllerClient.UnfenceClusterNetwork(ctx, &fenceRequest)
+	if err != nil {
+		klog.Errorf("Failed to unfence cluster network: %v", err)
+		return nil, err
+	}
+
+	return &proto.NetworkFenceResponse{}, nil
+}
+
+// getCIDRS converts the cidr string to a slice of cidrs.
+func getCIDRS(cidr []string) []*fence.CIDR {
```

Comment:
Make this method on the server struct ?

Reply:
why this is needed can you please explain more?

Reply:
> why this is needed can you please explain more?

it's just a thought, since it's only being used by method of that struct.

Reply:
generally, it's not a good idea to make methods if you are not using any member of the struct.